_posts/2020-04-09-flink-serialization-tuning-vol-1.md

```diff
@@ -0,0 +1,277 @@
+
+
+## PojoSerializer
+
+As outlined above, if your data type is not covered by a specialized serializer but follows the [POJO Rules](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/types_serialization.html#rules-for-pojo-types), it will be serialized with the [PojoSerializer](https://github.com/apache/flink/blob/release-1.10.0/flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java) which uses Java reflection to access an object’s fields. It is fast, generic, Flink-specific, and supports [state schema evolution](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/stream/state/schema_evolution.html) out of the box. If a composite data type cannot be serialized as a POJO, you will find the following message (or similar) in your cluster logs:
+
+> 15:45:51,460 INFO  org.apache.flink.api.java.typeutils.TypeExtractor             - Class … cannot be used as a POJO type because not all fields are valid POJO fields, and must be processed as GenericType. Please read the Flink documentation on "Data Types & Serialization" for details of the effect on performance.
+
+
+This means, that the PojoSerializer will not be used, but instead Flink will fall back to Kryo for serialization (see below). We will have a more detailed look into a few (more) situations that can lead to unexpected Kryo fallbacks in the second part of this blog post series.
+
+
+## Tuple Data Types
+
+Flink comes with a predefined set of tuple types which all have a fixed length and contain a set of strongly-typed fields of potentially different types. There are implementations for `Tuple0`, `Tuple1<T0>`, …, `Tuple25<T0, T1, ..., T24>` and they may serve as easy-to-use wrappers that spare the creation of POJOs for each and every combination of objects you need to pass between computations. With the exception of `Tuple0`*, these are serialized and deserialized with the [TupleSerializer](https://github.com/apache/flink/blob/release-1.10.0/flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/TupleSerializer.java) and the according fields’ serializers. Since tuple classes are completely under the control of Flink, both actions can be performed without reflection by accessing the appropriate fields directly. This certainly is a (performance) advantage when working with tuples instead of POJOs. Tuples, however, are not as flexible and certainly less descriptive in code.
```

Comment:
```suggestion
Flink comes with a predefined set of tuple types which all have a fixed length and contain a set of strongly-typed fields of potentially different types. There are implementations for `Tuple0`, `Tuple1<T0>`, …, `Tuple25<T0, T1, ..., T24>` and they may serve as easy-to-use wrappers that spare the creation of POJOs for each and every combination of objects you need to pass between computations. With the exception of `Tuple0`, these are serialized and deserialized with the [TupleSerializer](https://github.com/apache/flink/blob/release-1.10.0/flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/TupleSerializer.java) and the according fields’ serializers. Since tuple classes are completely under the control of Flink, both actions can be performed without reflection by accessing the appropriate fields directly. This certainly is a (performance) advantage when working with tuples instead of POJOs. Tuples, however, are not as flexible and certainly less descriptive in code.
```

Reply:
> With the exception of `Tuple0`*

Removed the *

Reply:
Yes, you're right...that's actually not needed with the info box so close. I also removed it from there.